### ISSUE TITLE
Construct regex with pattern of type [u8]?

### ISSUE THREAD

**Issue by iology** [rust-lang/regex#426]
Currently it only allows us to pass a UTF-8 encoded string.

**Comment by BurntSushi** [comment]
This isn't an actionable ticket.

1. What is your use case? What is the high level problem you're trying to solve?
2. Do you want to build a regex whose pattern has type `OsString`?
3. Do you want to match text whose type is `OsString`?
4. Both (2) and (3)?

**Comment by iology** [comment]
First, I know Unicode support cannot work well with byte strings.

My goal is to match byte strings, and build a regex from byte strings without resorting to escaping (not guranteed to work in anycase, like named captures). Also, extend this feature to `globset` so that I can build a glob with a byte string. All these is for my fork of *fd*, to match any file path without the limits of Unicode — https://github.com/jakwings/ff-find

(3) can be achieved by `regex::bytes::*`. So my aim is (2).

**Comment by BurntSushi** [comment]
OK, that clarifies some things, but I still don't understand why you want this. A regex like `r"(?-u)\xFF+"` is possible to write for example when using `regex::bytes`. Why is that insufficient?

> (not guranteed to work in all cases, like named captures)

Sorry, but I don't understand this. What isn't guaranteed to work?

**Comment by clarcharr** [comment]
I have a feeling that the crux of this would be having a regex search over a slice of `u16`, or potentially `u32`. Doing a search over `OsString` on most platforms would simply be a regex over bytes, but doing a search over `OsString` on Windows would require using half-words.

And it's really a question of whether this is necessary for this library. Maybe it might make sense to offer another library for it that reuses the internals from here.

**Comment by BurntSushi** [comment]
@clarcharr Well, this library certainly isn't going to support searching `u16` directly, but I feel like your comment made this ticket more confusing? It sounds like @iology wants to be able to do `let x: OsString = ...; let re = Regex::new(&x).unwrap()` where as you're talking about `let x: OsString = ...; let re = Regex::new("...").unwrap(); println!("{:?}", re.is_match(&x))`.

The simple work-around for your problem is to incur the performance hit and transcode to UTF-8 first. It's what ripgrep does, for example, on Windows.

**Comment by clarcharr** [comment]
Oh, I see, now.

The reason why I assumed the latter was because `bytes::Regex` scans bytes the constructor still takes a `str`. I imagined in this case that you'd construct something along the lines of an `os::Regex` with a `str` and scan `&OsStr` slices, giving `&OsStr` matches.

I completely forgot that on Windows, `OsStr` is re-encoded as WTF-8, which makes the whole `u16` scanning thing no longer relevant. That said, it might be worthwhile to add a shim over `bytes::Regex` to support scanning `OsStr`s, although I think that this wouldn't be worthwhile at this point considering how `OsStr` is generally designed to be passed around, not manipulated, and its current set of methods isn't very useful.

**Comment by iology** [comment]
Ah, I forgot that `as_bytes` is in the OsStrExt extension, so when I said `OsString` I meant `[u8]` (bytes).

There are other superset of ASCII like BIG5 and JIS. So I really want to remove the limits of UTF-8.

`regex-syntax` can help to escape invalid bytes but that's my last (hacky) resort. For me it is strange to have the UTF-8 limits.

If one happened to use non-ASCII text for the name of a capture group `(?P<æ>...)`, the error will be:

```
Error parsing regex near '3\xA6>...)' at character offset 12: Invalid capture name: '\xC3\xA6'. Capture names must consist of [_a-zA-Z0-9] and are not allowed to start with with a number.
```

Here `'\xC3\xA6'` come from my escaped input, so I cannot control this output style (unless I reparse the pattern with `regex-syntax`). More importantly, I cannot restore it to the original (unescaped) pattern string when other runtime errors occur.

Also I cannot know whether you'll implement something like "[Quoting escape](http://www.boost.org/doc/libs/1_65_1/libs/regex/doc/html/boost_regex/syntax/perl_syntax.html#boost_regex.syntax.perl_syntax.quoting_escape)" in the future, so I may finally use `regex-syntax`.

**Comment by BurntSushi** [comment]
> So I really want to remove the limits of UTF-8. [..] For me it is strange to have the UTF-8 limits.

UTF-8 is embedded all the way down to the lowest layers of the regex engine. So no matter what you do, you need to transcode to UTF-8 at some point. In theory, the regex parser could be modified to do this for you, but I haven't really been convinced. If you have a `&[u8]` regular expression, then you should be able to convert that to a `&str` regex by translating any non-UTF-8 bytes to escape sequences. Could you please explain why that's not sufficient? That's just what the parser would do anyway.

> If one happened to use non-ASCII text for the name of a capture group (?P<æ>...), the error will be:

This seems like a separate issue. Capture group names are limited to `[_0-9a-zA-Z]`. It might be better to expand that to `\pL` or something instead, but that's orthogonal from your other request of being able to use a `&[u8]` as a regex. Allow capture group names to truly be arbitrary bytes seems like a bad idea. Right now, they are intended to be limited to an indentifier-like alphabet.

**Comment by iology** [comment]
> Could you please explain why that's not sufficient?

Hmm, please reread my comment:

> More importantly, I cannot restore it to the original (unescaped) pattern string when other runtime errors occur.

The errors show the escaped string, not the original one. That's why I call it a dirty hack.

**Comment by BurntSushi** [comment]
Yeah I think you'll just need to translate the error message. I don't think it's worth changing the parser to support this case. I think I'd need to see compelling evidence that this isn't niche before there was enough motivation to do that.

**Comment by iology** [comment]
> … that this isn't niche

Yeah, I can understand how the evangelism-like ubiquity of `str` affect the refactoring work for this library. But don't think it is a niche, it is just that nobody want to put effort into this feature for now. Other non-rust libraries can do it, this library should have the ability to do it too.

> Yeah I think you'll just need to translate the error message.

Unfortunately that's impractical. It may show part of the pattern, so that I cann't get the exact locations.

**Comment by BurntSushi** [comment]
> Unfortunately that's impractical. It may show part of the pattern, so that I cann't get the exact locations.

Once the `regex-syntax` rewrite is done, you should be able to produce your own error message. The error type is exposed in `regex-syntax` today, but because of the [`surround` attribute](https://github.com/rust-lang/regex/blob/b9cb44733ccec141d43a302e8b27a66207e97e44/regex-syntax/src/lib.rs#L1352), you're hosed for the reason you stated. The rewrite won't have that problem.

> But don't think it is a niche, it is just that nobody want to put effort into this feature for now.

Umm, that's not what I meant. "Niche" means "not widely applicable." It's not just that nobody has wanted to put in the effort for this, nobody has voiced support for it at all. You're the first person to ask for it.

I still don't really understand the use case. Under what circumstances do you want to accept a regex in something other than UTF-8? Rust isn't really "evangelizing" as you say, UTF-8 is nearly ubiquitous outside of Rust. Does that mean there are zero use cases for non-UTF-8 encodings? No. Does that mean such use cases aren't important? Nope. It just means it's not something that comes up often, and that needs to be balanced against implementation effort.

Right now, the parser in regex-syntax (and the rewrite) are fundamentally based around parsing Unicode scalar values. What does changing that even mean? It certainly just can't be byte oriented, because you still want to support Unicode literal strings. So what do you need to do? You end up needing to support both a valid UTF-8 mode and an invalid-arbitrary-byte mode. That's an easy way to introduce bugs. That doesn't mean we should never do it, but it's implementation complexity, and that needs to be weighed against the utility of this feature you're requesting. It gets worse.

What happens when the regex bytes aren't ASCII-compatible or are encoded using a non-Unicode multi-byte encoding? So for example, if you had `[{xx}-{yy}]` where `{xx}` and `{yy}` are multi-byte chars, then what is the regex parser supposed to do with that? If it doesn't know about the encoding, then it can only operate on bytes, which will produce an unexpected result from the user's perspective. Otherwise, if it does know about the encoding, then your feature request evolves from "parse a regex in `&[u8]`" to "parse a `&[u8]` in any particular encoding that I specify." And now you're really screwed because the parser is still going to want a uniform character representation, and you've asked to introduced a new dependency in the regex crate for some arbitrary list of encodings.

> Other non-rust libraries can do it, this library should have the ability to do it too.

Which ones? Please show examples using those regex engines that satisfy your use case.

**Comment by iology** [comment]
I don't want to sounds like quarreling, and don't know why you asked me again and again. So I repeat:

> I still don't really understand the use case. Under what circumstances do you want to accept a regex in something other than UTF-8?

I've said it clearly from the start: When the (shell's) locale is BIG5, JIS, others. And make it easier to refactor `globset`: build a glob from a byte string.

> It just means it's not something that comes up often, and that needs to be balanced against implementation effort.

It doesn't come up often in the Rust land, because people tend to use `str` and that's what Rust recommends.

> That doesn't mean we should never do it, but it's implementation complexity, and that needs to be weighed against the utility of this feature you're requesting.

Yes, I can understand and I don't want to put big pressure on you. I may do a hard fork in the future but that's not ideal. So I first ask you here. **Please take it easy, ok?**

> What happens when the regex bytes aren't ASCII-compatible or are encoded using a non-Unicode multi-byte encoding?

I have said: superset of ASCII. Developers should read the documentation of regex carefully.

> So for example, if you had [{xx}-{yy}] where {xx} and {yy} are multi-byte chars, then what is the regex parser supposed to do with that? … Otherwise, if it does know about the encoding, then your feature request evolves from "parse a regex in &[u8]" to "parse a &[u8] in any particular encoding that I specify.

I do know that. And it is already the case when we use `regex::bytes::*`. It is no surprise. I also noted it in my *ff-find*. Bash's glob patterns also work this way.

> Please show examples using those regex engines that satisfy your use case.

```
echo $'\xC3' | LC_ALL=C grep -P $'\xC3'
```

**Comment by BurntSushi** [comment]
> I don't want to sounds like quarreling, and don't know why you asked me again and again. 

It's fine. I've just found it hard to understand what you were after here. This entire ticket has been very confusing. For example, I now realize that you aren't asking for sjis support (even though you mentioned it).

I do think I understand what you want now. I'm inclined to mark this as `wontfix`, unless someone can come up with a very simple way to handle this in the parser.

**Comment by iology** [comment]
Sure, if you pleased. I don't understand what `wonfix` means while if someone can do it. Let me close it by myself.
